File: src/main/java/org/tobi29/scapes/engine/gui/GuiUtils.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.engine.gui;

import org.tobi29.scapes.engine.graphics.Mesh;
import org.tobi29.scapes.engine.utils.math.vector.Vector3;

public final class GuiUtils {
    private GuiUtils() {
    }

    public static void rectangle(GuiRenderBatch renderer, float minX,
            float minY, float maxX, float maxY, float r, float g, float b,
            float a) {
        rectangle(renderer, minX, minY, maxX, maxY, 0.0f, 0.0f, 1.0f, 1.0f, r,
                g, b, a);
    }

    public static void rectangle(GuiRenderBatch renderer, float minX,
            float minY, float maxX, float maxY, float minTX, float minTY,
            float maxTX, float maxTY, float r, float g, float b, float a) {
        Mesh mesh = renderer.mesh();
        Vector3 tl = renderer.vector(minX, minY);
        Vector3 tr = renderer.vector(maxX, minY);
        Vector3 bl = renderer.vector(minX, maxY);
        Vector3 br = renderer.vector(maxX, maxY);
        mesh.addVertex(tl.floatX(), tl.floatY(), 0.0f, r, g, b, a, minTX,
                minTY);
        mesh.addVertex(tr.floatX(), tr.floatY(), 0.0f, r, g, b, a, maxTX,
                minTY);
        mesh.addVertex(bl.floatX(), bl.floatY(), 0.0f, r, g, b, a, minTX,
                maxTY);
        mesh.addVertex(bl.floatX(), bl.floatY(), 0.0f, r, g, b, a, minTX,
                maxTY);
        mesh.addVertex(tr.floatX(), tr.floatY(), 0.0f, r, g, b, a, maxTX,
                minTY);
        mesh.addVertex(br.floatX(), br.floatY(), 0.0f, r, g, b, a, maxTX,
                maxTY);
    }

    public static void shadow(GuiRenderBatch renderer, float minX, float minY,
            float maxX, float maxY, float r, float g, float b, float a) {
        shadow(renderer, minX, minY, maxX, maxY, 8.0f, r, g, b, a);
    }

    public static void shadow(GuiRenderBatch renderer, float minX, float minY,
            float maxX, float maxY, float radius, float r, float g, float b,
            float a) {
        Mesh mesh = renderer.mesh();
        Vector3 tli = renderer.vector(minX, minY);
        Vector3 tri = renderer.vector(maxX, minY);
        Vector3 bli = renderer.vector(minX, maxY);
        Vector3 bri = renderer.vector(maxX, maxY);
        Vector3 tlt = renderer.vector(minX, minY - radius);
        Vector3 trt = renderer.vector(maxX, minY - radius);
        Vector3 blb = renderer.vector(minX, maxY + radius);
        Vector3 brb = renderer.vector(maxX, maxY + radius);
        Vector3 tll = renderer.vector(minX - radius, minY);
        Vector3 trr = renderer.vector(maxX + radius, minY);
        Vector3 bll = renderer.vector(minX - radius, maxY);
        Vector3 brr = renderer.vector(maxX + radius, maxY);
        // Top
        mesh.addVertex(tlt.floatX(), tlt.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(trt.floatX(), trt.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(tli.floatX(), tli.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(tli.floatX(), tli.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(trt.floatX(), trt.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(tri.floatX(), tri.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        // Bottom
        mesh.addVertex(bli.floatX(), bli.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(bri.floatX(), bri.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(blb.floatX(), blb.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(blb.floatX(), blb.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(bri.floatX(), bri.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(brb.floatX(), brb.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        // Left
        mesh.addVertex(tll.floatX(), tll.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(tli.floatX(), tli.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(bll.floatX(), bll.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(bll.floatX(), bll.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(tli.floatX(), tli.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(bli.floatX(), bli.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        // Right
        mesh.addVertex(tri.floatX(), tri.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(trr.floatX(), trr.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(bri.floatX(), bri.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(bri.floatX(), bri.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(trr.floatX(), trr.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(brr.floatX(), brr.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        // Top-left
        mesh.addVertex(tli.floatX(), tli.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(tll.floatX(), tll.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(tlt.floatX(), tlt.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        // Top-right
        mesh.addVertex(tri.floatX(), tri.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(trt.floatX(), trt.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(trr.floatX(), trr.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        // Bottom-left
        mesh.addVertex(bli.floatX(), bli.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(blb.floatX(), blb.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(bll.floatX(), bll.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        // Bottom-right
        mesh.addVertex(bri.floatX(), bri.floatY(), 0.0f, r, g, b, a, 0.0f,
                0.0f);
        mesh.addVertex(brr.floatX(), brr.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
        mesh.addVertex(brb.floatX(), brb.floatY(), 0.0f, r, g, b, 0.0f, 0.0f,
                0.0f);
    }
}

File: src/main/kotlin/org/tobi29/scapes/engine/gui/GuiUtils.kt
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.gui

object GuiUtils {

    fun rectangle(renderer: GuiRenderBatch,
                  minX: Float,
                  minY: Float,
                  maxX: Float,
                  maxY: Float,
                  r: Float,
                  g: Float,
                  b: Float,
                  a: Float) {
        rectangle(renderer, minX, minY, maxX, maxY, 0.0f, 0.0f, 1.0f, 1.0f, r,
                g, b, a)
    }

    fun rectangle(renderer: GuiRenderBatch,
                  minX: Float,
                  minY: Float,
                  maxX: Float,
                  maxY: Float,
                  minTX: Float,
                  minTY: Float,
                  maxTX: Float,
                  maxTY: Float,
                  r: Float,
                  g: Float,
                  b: Float,
                  a: Float) {
        val mesh = renderer.mesh()
        val tl = renderer.vector(minX.toDouble(), minY.toDouble())
        val tr = renderer.vector(maxX.toDouble(), minY.toDouble())
        val bl = renderer.vector(minX.toDouble(), maxY.toDouble())
        val br = renderer.vector(maxX.toDouble(), maxY.toDouble())
        mesh.addVertex(tl.floatX().toDouble(), tl.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(),
                minTX.toDouble(),
                minTY.toDouble())
        mesh.addVertex(tr.floatX().toDouble(), tr.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(),
                maxTX.toDouble(),
                minTY.toDouble())
        mesh.addVertex(bl.floatX().toDouble(), bl.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(),
                minTX.toDouble(),
                maxTY.toDouble())
        mesh.addVertex(bl.floatX().toDouble(), bl.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(),
                minTX.toDouble(),
                maxTY.toDouble())
        mesh.addVertex(tr.floatX().toDouble(), tr.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(),
                maxTX.toDouble(),
                minTY.toDouble())
        mesh.addVertex(br.floatX().toDouble(), br.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(),
                maxTX.toDouble(),
                maxTY.toDouble())
    }

    fun shadow(renderer: GuiRenderBatch,
               minX: Float,
               minY: Float,
               maxX: Float,
               maxY: Float,
               r: Float,
               g: Float,
               b: Float,
               a: Float) {
        shadow(renderer, minX, minY, maxX, maxY, 8.0f, r, g, b, a)
    }

    fun shadow(renderer: GuiRenderBatch,
               minX: Float,
               minY: Float,
               maxX: Float,
               maxY: Float,
               radius: Float,
               r: Float,
               g: Float,
               b: Float,
               a: Float) {
        val mesh = renderer.mesh()
        val tli = renderer.vector(minX.toDouble(), minY.toDouble())
        val tri = renderer.vector(maxX.toDouble(), minY.toDouble())
        val bli = renderer.vector(minX.toDouble(), maxY.toDouble())
        val bri = renderer.vector(maxX.toDouble(), maxY.toDouble())
        val tlt = renderer.vector(minX.toDouble(), (minY - radius).toDouble())
        val trt = renderer.vector(maxX.toDouble(), (minY - radius).toDouble())
        val blb = renderer.vector(minX.toDouble(), (maxY + radius).toDouble())
        val brb = renderer.vector(maxX.toDouble(), (maxY + radius).toDouble())
        val tll = renderer.vector((minX - radius).toDouble(), minY.toDouble())
        val trr = renderer.vector((maxX + radius).toDouble(), minY.toDouble())
        val bll = renderer.vector((minX - radius).toDouble(), maxY.toDouble())
        val brr = renderer.vector((maxX + radius).toDouble(), maxY.toDouble())
        // Top
        mesh.addVertex(tlt.floatX().toDouble(), tlt.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(trt.floatX().toDouble(), trt.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(tli.floatX().toDouble(), tli.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(tli.floatX().toDouble(), tli.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(trt.floatX().toDouble(), trt.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(tri.floatX().toDouble(), tri.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        // Bottom
        mesh.addVertex(bli.floatX().toDouble(), bli.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(bri.floatX().toDouble(), bri.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(blb.floatX().toDouble(), blb.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(blb.floatX().toDouble(), blb.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(bri.floatX().toDouble(), bri.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(brb.floatX().toDouble(), brb.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        // Left
        mesh.addVertex(tll.floatX().toDouble(), tll.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(tli.floatX().toDouble(), tli.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(bll.floatX().toDouble(), bll.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(bll.floatX().toDouble(), bll.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(tli.floatX().toDouble(), tli.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(bli.floatX().toDouble(), bli.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        // Right
        mesh.addVertex(tri.floatX().toDouble(), tri.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(trr.floatX().toDouble(), trr.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(bri.floatX().toDouble(), bri.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(bri.floatX().toDouble(), bri.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(trr.floatX().toDouble(), trr.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(brr.floatX().toDouble(), brr.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        // Top-left
        mesh.addVertex(tli.floatX().toDouble(), tli.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(tll.floatX().toDouble(), tll.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(tlt.floatX().toDouble(), tlt.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        // Top-right
        mesh.addVertex(tri.floatX().toDouble(), tri.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(trt.floatX().toDouble(), trt.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(trr.floatX().toDouble(), trr.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        // Bottom-left
        mesh.addVertex(bli.floatX().toDouble(), bli.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(blb.floatX().toDouble(), blb.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(bll.floatX().toDouble(), bll.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        // Bottom-right
        mesh.addVertex(bri.floatX().toDouble(), bri.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), a.toDouble(), 0.0,
                0.0)
        mesh.addVertex(brr.floatX().toDouble(), brr.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
        mesh.addVertex(brb.floatX().toDouble(), brb.floatY().toDouble(), 0.0,
                r.toDouble(), g.toDouble(), b.toDouble(), 0.0, 0.0,
                0.0)
    }
}
